darwin/importer/importer.py

```diff
@@ -0,0 +1,106 @@
+from pathlib import Path
+from typing import Callable, List, Optional, Union
+
+import darwin.datatypes as dt
+from darwin.utils import secure_continue_request
+
+# def make_polygon(class_name, point_path):
+#     return Annotation(AnnotationClass(class_name, "polygon"), {"path": point_path})
+
+
+def build_main_annotations_lookup_table(annotation_classes):
+    lookup = {}
+    for cls in annotation_classes:
+        for annotation_type in cls["annotation_types"]:
+            if annotation_type["granularity"] == "main":
+                if annotation_type["name"] not in lookup:
+                    lookup[annotation_type["name"]] = {}
+
+                lookup[annotation_type["name"]][cls["name"]] = cls["id"]
+    return lookup
+
+
+def import_annotations(
+    dataset: "RemoteDataset",
+    importer: Callable[[Path], Optional[dt.AnnotationFile]],
+    file_paths: List[Union[str, Path]],
+):
+    print("Fetching remote file list...")
+    remote_files = {f["filename"]: f["id"] for f in dataset.fetch_remote_files()}
+    print("Fetching remote class list...")
+    remote_classes = build_main_annotations_lookup_table(dataset.fetch_remote_classes())
+
+    print("Retrieving local annotations ...")
+    local_files = []
+    local_files_missing_remotely = []
+    # TODO: this could be done in parallell
+    for file_path in map(Path, file_paths):
+        files = file_path.glob("**/*") if file_path.is_dir() else [file_path]
+        for f in files:
+            parsed_file = importer(f)
+            if not parsed_file:
+                continue
+            # clear to save memory
+            parsed_file.annotations = []
+            if parsed_file.filename not in remote_files:
+                local_files_missing_remotely.append(parsed_file)
+                continue
+            local_files.append(parsed_file)
```

Comment:
I would also maybe extract this, the `import_annotations` function is rather large otherwise